code/Model/SiteTree.php

```diff
@@ -2959,17 +2977,28 @@ public function getControllerName()
             }
             // If we have a class of "{$ClassName}Controller" then we found our controller
             if (class_exists($candidate = sprintf('%sController', $class))) {
-                $controller = $candidate;
-                break;
+                return $candidate;
             } elseif (class_exists($candidate = sprintf('%s_Controller', $class))) {
                 // Support the legacy underscored filename, but raise a deprecation notice
                 Deprecation::notice(
                     '5.0',
                     'Underscored controller class names are deprecated. Use "MyController" instead of "My_Controller".',
                     Deprecation::SCOPE_GLOBAL
                 );
-                $controller = $candidate;
-                break;
+                return $candidate;
+            } elseif (is_array($namespaceMap)) {
+                foreach ($namespaceMap as $pageNamespace => $controllerNamespace) {
+                    if (strpos($class, $pageNamespace) === false) {
```

Comment:
```suggestion
                    if (strpos($class, $pageNamespace) !== 0) {
```
Bit of an edge case but we should probably make sure the mapping namespace isn't actually somewhere in the middle of the class name.